### PR TITLE
Use SiteURL for all calls from the webapp

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
Use SiteURL for all calls from the webapp

#### Ticket Link
https://github.com/mattermost/mattermost-plugin-skype4business/issues/61

### CODE REVIEW

**Review by larkox — APPROVED** [review]

File: webapp/src/client/client.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ import AuthenticationContext from 'adal-angular';
 import {isDesktopApp} from '../utils/user_utils';
 import {Periods} from '../constants';
 import {id as pluginID} from '../manifest';
+import { getPluginServerRoute } from 'selectors';
```
Comment (mickmister) [inline webapp/src/client/client.js:8]: To stay consistent with the other imports

```suggestion
import {getPluginServerRoute} from 'selectors';
```

File: webapp/src/client/client.js [hunk 2]
```diff
@@ -5,6 +5,8 @@ import AuthenticationContext from 'adal-angular';
...
+
```
Comment (mickmister) [inline webapp/src/client/client.js:9]: Line 9 can be removed

File: webapp/src/client/client.js [hunk 3]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
     };
 
-    doGet = async (url, headers = {}, credentials) => {
-        headers.Accept = 'application/json';
-        let options = {
-            method: 'get',
-            headers,
-        };
+    
+    doGet = (url, headers = {}, credentials) => {
```
Comment (mickmister) [inline webapp/src/client/client.js:270]: Line 268 can be removed

File: webapp/src/client/client.js [hunk 4]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
+            let options = {
+                method: 'get',
+                headers,
+            };
 
-        if (credentials) {
-            options.credentials = credentials;
-        }
+            if (url.includes('plugins/' + pluginID)) {
+                options = Client4.getOptions(options);
```
Comment (mickmister) [inline webapp/src/client/client.js:280]: The prefix of the server URL should only occur on requests going to the Mattermost server. In this case all of the Mattermost requests include this plugin id structure. I think this makes sense:

```suggestion
                url = getPluginServerRoute(getState()) + "/"+ url;
                options = Client4.getOptions(options);
```

Then we can remove line 272

File: webapp/src/client/client.js [hunk 5]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
+                 headers,
+             };
 
-        throw new ClientError(Client4.url, {
-            message: text || '',
-            status_code: response.status,
-            url,
-        });
+             if (url.includes('plugins/' + pluginID)) {
+                 options = Client4.getOptions(options);
```
Comment (mickmister) [inline webapp/src/client/client.js:321]: Same suggestion here regarding applying the server URL to Mattermost-specific URLs only:

```suggestion
                 url = getPluginServerRoute(getState()) + "/"+ url;
                 options = Client4.getOptions(options);
```

Then we can remove line 311

File: webapp/src/client/client.js [hunk 6]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
+
+             const text = await response.text();
+
+             throw new ClientError(Client4.url, {
+                 message: text || '',
+                 status_code: response.status,
+                 url,
+             });
+
+            //return true;
```
Comment (mickmister) [inline webapp/src/client/client.js:338]: Lines 337 and 338 can be removed

File: webapp/src/client/client.js [hunk 7]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
+        };
+
```
Comment (mickmister) [inline webapp/src/client/client.js:340]: Line 340 can be removed

File: webapp/src/client/client.js [hunk 8]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
-        if (url.includes('plugins/' + pluginID)) {
-            options = Client4.getOptions(options);
-        }
 
-        const response = await fetch(url, options);
 
-        if (response.ok) {
-            return response.json();
-        }
+
```
Comment (mickmister) [inline webapp/src/client/client.js:306]: This line and  the other empty lines above can be removed

File: webapp/src/selectors/index.js [hunk 9]
```diff
@@ -1,10 +1,29 @@
 import {id as pluginId} from '../manifest';
 
+import { getConfig } from 'mattermost-redux/selectors/entities/general';
```
Comment (mickmister) [inline webapp/src/selectors/index.js:3]:
```suggestion
import {getConfig} from 'mattermost-redux/selectors/entities/general';
```

File: webapp/src/selectors/index.js [hunk 10]
```diff
@@ -1,10 +1,29 @@
...
+    let basePath = '';
+    if (config && config.SiteURL) {
+        basePath = new URL(config.SiteURL).pathname;
+
+        if (basePath && basePath[basePath.length - 1] === '/') {
+            basePath = basePath.substr(0, basePath.length - 1);
+        }
+    }
+
+    // return basePath + '/plugins/' + PluginId;
```
Comment (mickmister) [inline webapp/src/selectors/index.js:21]: Line 21 can be removed

File: webapp/src/selectors/index.js [hunk 11]
```diff
@@ -1,10 +1,29 @@
...
+    return basePath;
+};
+
 export default {
     getAuthenticationResult,
 };
 
+
```
Comment (mickmister) [inline webapp/src/selectors/index.js:29]: This line can be removed

File: webapp/src/client/client.js [hunk 12]
```diff
@@ -263,62 +265,79 @@ export default class Client {
...
+            });
 
-    doPost = async (url, body, headers = {}) => {
-        headers.Accept = 'application/json';
-        headers['Content-Type'] = 'application/json';
-        let options = {
-            method: 'post',
-            body: JSON.stringify(body),
-            headers,
+            //return true;
```
Comment (mickmister) [inline webapp/src/client/client.js:301]: This line and the one above it can be removed

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks for implementing these changes @maisnamrajusingh :+1:

I have a few requests regarding styling of imports and removing empty whitespace lines, and prepending the server URL only when the request is for the Mattermost server.

**Comment by mickmister** [comment]
@maisnamrajusingh Were you able to test this code using a skype server

**Comment by mickmister** [comment]
There are actually no calls to the Mattermost server with the client at `webapp/src/client/client.js`, so we actually never want to use the site URL. I propose to:
- revert all changes in this PR
- rename this `client.js` file to `skype_api.js` and remove the `client/index.js` file
- remove any calls to `Client4`, including `Client4.getOptions`

**Comment by maisnamraju** [comment]
@mickmister  I remember I had setup a skype business account for this but let me confirm if this works by testing it tomorrow and fix the ci issues.

File: webapp/src/client/client.js [hunk 13]
```diff
@@ -1,20 +1,21 @@
...
+import {ClientError} from "mattermost-redux/client/client4";
+import AuthenticationContext from "adal-angular";
 
-import {isDesktopApp} from '../utils/user_utils';
-import {Periods} from '../constants';
-import {id as pluginID} from '../manifest';
+import {isDesktopApp} from "../utils/user_utils";
+import {Periods} from "../constants";
+import {id as pluginID} from "../manifest";
+import {getPluginServerRoute} from "selectors";
```
Comment (hanzei) [inline webapp/src/client/client.js:8]: Single quotes are the de-fact standard for js code at Mattermost. Why does this PR change them?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: webapp/src/client/client.js (hunk 13 above)
Comment (maisnamrajusingh) [reply webapp/src/client/client.js:8]: @hanzei  sorry about this. This was one of the first PRs that got sent out and this has been ignored for a while. I'll pick this up and clean it up

File: webapp/src/client/client.js (hunk 1 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js]: @mickmister This is Done

File: webapp/src/client/client.js (hunk 2 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js]: done

File: webapp/src/client/client.js (hunk 3 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js]: done

File: webapp/src/client/client.js (hunk 13 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js]: @hanzei  removed this unwanted changes

File: webapp/src/client/client.js [hunk 14]
```diff
@@ -263,68 +264,76 @@ export default class Client {
...
-
-        if (url.includes('plugins/' + pluginID)) {
-            options = Client4.getOptions(options);
-        }
-
-        if (credentials) {
-            options.credentials = credentials;
-        }
+    doGet = (url, headers = {}, credentials) => {
+        return async (dispatch, getState) => {
```
Comment (mickmister) [inline webapp/src/client/client.js:268]: Don't we need to `dispatch` these functions since they are returning [thunks](https://github.com/reduxjs/redux-thunk#motivation) now?

As in, anywhere we are calling `doGet(url)`, we need to change this to be `dispatch(doGet(url))`. 1/5 Maybe we should go the route of "setting" the site URL on the client as we have done in similar PRs for other plugins.

File: webapp/src/client/client.js [hunk 15]
```diff
@@ -263,68 +264,76 @@ export default class Client {
...
-            status_code: response.status,
-            url,
-        });
+            throw new ClientError(Client4.url, {
+                message: text || '',
+                status_code: response.status,
+                url,
+            });
+
+            // return true;
```
Comment (mickmister) [inline webapp/src/client/client.js:329]: Looks like extra code got left over here

File: webapp/src/client/client.js [hunk 16]
```diff
@@ -263,68 +264,76 @@ export default class Client {
...
+        };
     };
 
     generateUuid4 = () => {
         return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
             // eslint-disable-next-line
-            let r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8);
+            let r = (Math.random() * 16) | 0,v = c === 'x' ? r : (r & 0x3) | 0x8;
```
Comment (mickmister) [inline webapp/src/client/client.js:336]: Seems like an unnecessary change

File: webapp/src/selectors/index.js [hunk 17]
```diff
@@ -1,10 +1,24 @@
...
+        if (basePath && basePath[basePath.length - 1] === '/') {
+            basePath = basePath.substr(0, basePath.length - 1);
+        }
+    }
+
+    return basePath;
 };
 
+export default {getAuthenticationResult};
```
Comment (mickmister) [inline webapp/src/selectors/index.js:24]: Can we instead export `getAuthenticationResult` like below, and adjust imports of the function to expect it to be exported in this manner?

```suggestion
export const getAuthenticationResult = (state) => getPluginState(state).authenticationResult;
```

File: webapp/src/client/client.js [hunk 18]
```diff
@@ -5,6 +5,7 @@ import AuthenticationContext from 'adal-angular';
 import {isDesktopApp} from '../utils/user_utils';
 import {Periods} from '../constants';
 import {id as pluginID} from '../manifest';
+import {getPluginServerRoute} from 'selectors';
```
Comment (mickmister) [inline webapp/src/client/client.js:8]: Can we make this a relative import `../selectors` like the other local imports here?

File: webapp/src/client/client.js [hunk 19]
```diff
@@ -263,68 +264,76 @@ export default class Client {
...
-        if (url.includes('plugins/' + pluginID)) {
-            options = Client4.getOptions(options);
-        }
-
-        if (credentials) {
-            options.credentials = credentials;
-        }
+    doGet = (url, headers = {}, credentials) => {
+        return async (dispatch, getState) => {
+            const baseUrl = getPluginServerRoute(getState()) + '/' + url;
```
Comment (mickmister) [inline webapp/src/client/client.js:269]: We want to make sure we're only adjusting the URL if the call is to the Mattermost server. Is it possible `doGet` or `doPost` is ever called with a URL that is pointing to a Skype server? In this case we would not want to prepend the site URL.

**Review by mickmister — CHANGES_REQUESTED** [review]

File: webapp/src/client/client.js (hunk 19 above)
Comment (mickmister) [reply webapp/src/client/client.js:269]: This was also brought up in https://github.com/mattermost/mattermost-plugin-skype4business/pull/78#discussion_r669253152

Comment (mickmister) [reply webapp/src/client/client.js:269]: Also `baseUrl` seems like the wrong name for this variable, since this is the URL we will ultimately use in the request.

```suggestion
            url = getPluginServerRoute(getState()) + '/' + url;
```

File: webapp/src/client/client.js (hunk 14 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js:268]: @mickmister I agree with what you mean regards to the setup of the site url but the implementation is different on many of the plugins, can you tell me which plugin I should follow

File: webapp/src/selectors/index.js (hunk 17 above)
Comment (sibasankarnayak) [reply webapp/src/selectors/index.js:24]: Removed

File: webapp/src/client/client.js (hunk 18 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js:8]: Removed

File: webapp/src/client/client.js (hunk 16 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js:336]: Removed

File: webapp/src/client/client.js (hunk 15 above)
Comment (sibasankarnayak) [reply webapp/src/client/client.js:329]: Removed

File: webapp/src/client/client.js (hunk 14 above)
Comment (mickmister) [reply webapp/src/client/client.js]: Let's go with the implementation from https://github.com/mattermost/mattermost-plugin-github/pull/459/files

File: webapp/src/selectors/index.js [hunk 20]
```diff
@@ -1,10 +1,21 @@
...
+        basePath = new URL(config.SiteURL).pathname;
 
+        if (basePath && basePath[basePath.length - 1] === '/') {
+            basePath = basePath.substr(0, basePath.length - 1);
+        }
+    }
+
+    return basePath;
+};
+export const getAuthenticationResult = (state) => getPluginState(state).authenticationResult;
```
Comment (mickmister) [inline webapp/src/selectors/index.js:21]:
```suggestion
};

export const getAuthenticationResult = (state) => getPluginState(state).authenticationResult;
```

**Comment by mickmister** [comment]
This looks good other than the two remaining concerns https://github.com/mattermost/mattermost-plugin-skype4business/pull/78#discussion_r720376046 (avoiding prepending site url to requests going to skype server) and https://github.com/mattermost/mattermost-plugin-skype4business/pull/78#discussion_r722327597 (set the client URL on plugin start up instead of using thunks within the client). Thanks @sibasankarnayak :+1:
